### PR TITLE
ARROW-11412: [Python] Improve Expression docs

### PR DESCRIPTION
**Pull request by romankarlstetter**
Slightly improve documentation on how to create and combine expressions
for filtering datasets. Including one example.

### CODE REVIEW

File: python/pyarrow/_dataset.pyx [hunk 1]
```diff
@@ -84,7 +84,25 @@ cdef CFileSource _make_file_source(object file, FileSystem filesystem=None):
 
 
 cdef class Expression(_Weakrefable):
-
+    """
+    A logical expression to be evaluated against some input.
+
+    To create an expression:
+     - Use the factory function ``pyarrow.dataset.scalar()`` to create a
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:91]: Can you leave a blank line above this one (to separate the text from the list)?

And the `-` has one extra space before it as indentation, which needs to be removed in order to render it correctly with sphinx.

(restructured text is quite picky ... ;))

File: python/pyarrow/_dataset.pyx [hunk 2]
```diff
@@ -84,7 +84,25 @@ cdef CFileSource _make_file_source(object file, FileSystem filesystem=None):
...
+     - Use the factory function ``pyarrow.dataset.field()`` to reference
+       a field (column in table).
+     - Compare fields and scalars with ``<``, ``<=``, ``==``, ``>=``, ``>``.
+     - Combine expressions using python operators ``&`` (logical and),
+       ``|`` (logical or) and ``~`` (logical not).
+       Note: python keywords ``and``, ``or`` and ``not`` do not work to combine expressions.
+
+    Examples:
+    --------
+    >>> from pyarrow.dataset import scalar, field
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:102]: In the other part of the docs, we use `import pyarrow.dataset as ds`, we should probably be consistent (https://arrow.apache.org/docs/python/dataset.html)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot! 
(yeah, the docstrings are quite lacking ..)

Just a few formatting comments.

**Comment by romankarlstetter** [comment]
I tried to fix the issues you mentioned and also updated dataset.rst to clarify the use of &, | ~ vs and or not

**Review by jorisvandenbossche — APPROVED** [review]
Thanks for the updates!

**Pull request closed without merging**